### PR TITLE
das: unify logic in catchUp and sample

### LINKED ISSUE #801: das: Simplify `catchUp` and `sample` routines to rely on the same logic
Deduplicate the current logic inside of `daser.go` such that both the `catchUp` and `sample` loops rely on the same logic.

### PR DESCRIPTION
**Pull request by distractedm1nd**
Would close #801 . @renaynay Is this what you meant, or did you have something else in mind?

When I read the issue I assumed it was about the entire loop logic, but it seems that abstracting it would complicate both routines instead of making it more readable.

### CODE REVIEW

**Comment by renaynay** [comment]
yes @distractedm1nd this is what i had in mind. from a quick skim, looks good. Will review in depth tomorrow. thanks!

File: das/daser.go [hunk 1]
```diff
@@ -220,28 +204,38 @@ func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) (int64, error) {
...
+		log.Warn("DASer WILL BE STOPPED. IN ORDER TO CONTINUE SAMPLING, RE-START THE NODE")
+		// report previous height as the last successfully sampled height
+		return h.Height, err
+	}
+
+	sampleTime := time.Since(startTime)
+	log.Infow("sampled header", "height", h.Height, "hash", h.Hash(),
+		"square width", len(h.DAH.RowsRoots), "finished (s)", sampleTime.Seconds())
+
+	return h.Height, nil
```
Comment (adlerjohn) [inline das/daser.go:240]: Why does this function need to return height at all? It always returns `h.Height` in all execution paths.

Comment (distractedm1nd) [reply das/daser.go]: True, didn't catch that - resolved now

File: das/daser.go [hunk 2]
```diff
@@ -136,23 +136,7 @@ func (d *DASer) sample(ctx context.Context, sub header.Subscription, checkpoint
...
-				"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "err", err)
-			log.Warn("DASer WILL BE STOPPED. IN ORDER TO CONTINUE SAMPLING, RE-START THE NODE")
-			return
-		}
-
-		sampleTime := time.Since(startTime)
-		log.Infow("sampling successful", "height", h.Height, "hash", h.Hash(),
-			"square width", len(h.DAH.RowsRoots), "finished (s)", sampleTime.Seconds())
-
+		d.sampleHeader(ctx, h)
```
[the remark this answers is not in the document]
Comment (distractedm1nd) [reply das/daser.go]: Fixed - thanks. My IDE normally catches this as well, not sure how I missed it. Is there a linter in the CI that you mean? It probably just hasn't ran yet, it needs approval

**Review by adlerjohn — APPROVED** [review]

File: das/daser.go [hunk 3]
```diff
@@ -220,28 +207,36 @@ func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) (int64, error) {
...
+
+	err := d.da.SharesAvailable(ctx, h.DAH)
+	if err != nil {
+		if err == context.Canceled {
+			return nil
+		}
+		log.Errorw("sampling failed", "height", h.Height, "hash", h.Hash(),
+			"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "err", err)
+		log.Warn("DASer WILL BE STOPPED. IN ORDER TO CONTINUE SAMPLING, RE-START THE NODE")
+		// report previous height as the last successfully sampled height
```
Comment (Wondertan) [inline das/daser.go:233]: This comment is no more relevant here. Pls move it back.

File: das/daser.go [hunk 4]
```diff
@@ -220,28 +207,36 @@ func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) (int64, error) {
...
+	startTime := time.Now()
+
+	err := d.da.SharesAvailable(ctx, h.DAH)
+	if err != nil {
+		if err == context.Canceled {
+			return nil
+		}
+		log.Errorw("sampling failed", "height", h.Height, "hash", h.Hash(),
+			"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "err", err)
+		log.Warn("DASer WILL BE STOPPED. IN ORDER TO CONTINUE SAMPLING, RE-START THE NODE")
```
Comment (Wondertan) [inline das/daser.go:232]: This log warn is only relevant to errored `sampleHeader` call in `sample` method. Put it to loc 141.

**Comment by Wondertan** [comment]
@distractedm1nd, hi, this is a good PR and we would like to merge it before another opened one #820, so it's kinda of blocking already. Can you pls resolve open comments ASAP so we can proceed?

**Comment by liamsi** [comment]
@Wondertan, alternatively, merge it as is, and address the open comments in a followup PR afterwards.

**Review by Wondertan — APPROVED** [review]

**Pull request merged**